caffe2/operators/conv_gradient_op.cc

```diff
@@ -4,8 +4,85 @@
 
 namespace caffe2 {
 
+std::vector<TensorShape> TensorInferenceForConvGradient(
+    const OperatorDef& def,
+    const std::vector<TensorShape>& in) {
+  CAFFE_ENFORCE(in.size() >= 2, "ConvGradient expects at least 2 inputs");
+
+  if (in[0].unknown_shape()) {
+    std::vector<TensorShape> out(1);
+    out[0].set_unknown_shape(true);
+    return out;
+  }
+  ArgumentHelper helper(def);
+  const auto no_bias = helper.GetSingleArgument<int>("no_bias", 0);
+  const auto n_outputs = def.output_size();
+  auto i = 0;
+  vector<TensorShape> out(n_outputs);
+
+  out[i++] = in[1];
+  if (!no_bias) {
+    vector<int> bias_shape = {in[1].dims(0)};
+    out[i++] = CreateTensorShape(bias_shape, in[1].data_type());
+  }
+
+  if (n_outputs == 3 || (no_bias && n_outputs == 2)) {
+    out[i++] = in[0];
+  }
+
+  return out;
+}
+
+OpSchema::Cost CostInferenceForConvGradient(
+    const OperatorDef& def,
+    const vector<TensorShape>& inputs) {
+  CAFFE_ENFORCE_EQ(inputs.size(), 3, "ConvGradient requires at least 3 inputs");
+  ArgumentHelper helper(def);
+  const auto order =
+      StringToStorageOrder(helper.GetSingleArgument<string>("order", "NCHW"));
+  const auto no_bias = helper.GetSingleArgument<int>("no_bias", 0);
+  const auto n_outputs = def.output_size();
+
+  const auto& outputs = TensorInferenceForConvGradient(def, inputs);
+  const auto& X = inputs[0];
+  const auto& filter = inputs[1];
+  const auto& dY = inputs[2];
+  const auto N = X.dims(0);
+  const auto M = filter.dims(0);
+  const auto C = (order == StorageOrder::NCHW ? X.dims(1) : X.dims(3));
```

Comment:
@BIT-silence is working on 3D conv in NHWC so let's make it general.
``
  const auto C = (order == StorageOrder::NCHW ? X.dims(1) : X.dims(X.dims_size() - 1));
``